Search improvements

silas opened the pull request:
Some of the things I've changed/tweaked:
 * Use one index instead of two
 * Attempted to improve ranking/scoring
 * Truncate summaries while attempting to retain useful highlights
 * Fix #2438
 * Remove results that point to the same anchor

I've done a bunch of manual side-by-side comparison of the results on the Nextra docs site, the swr-site, and my private project.

I've personally found the results to be much better with the new search logic, but I'd love to get another set of eyes on it, and any example of regressions so I can dig into them.

You added a test for no results for `>  a` which I purposely didn't address. I'm not sure it's a useful issue to fix and it's going to do weird stuff for people entering Chinese/Japanese/Korean characters (without adding language aware tokenizers). The current code handles boosting the score for these languages pretty poorly, but I didn't want to just hard filter them out.

Might be nice to get the preview version of this deployed to vercel to make the side-by-side comparison easier (assuming v3 branch is deployed somewhere we can compare it to).

---

Things that probably need more work/resarch:
 * ~Summaries aren't truncated at word boundaries~
 * ~Titles aren't highlighted~
 * ~Tests~
 * ~Make sure memory usage didn't get worse~

From a UI perspective it might be worth considering adding page links and making the page title/item selectable, which would alleviate the need for the placeholder section item.

---

Things to follow up on later:

 * look into including page hierarchy in the index/title (e.g. I have a couple of "Getting Started" guides in my docs and it's difficult to tell them apart in the search). Having the context would also improve the ranking of results
 * dig into the `structurizedData` data generation and attempt to fix merged paragraphs
 * look into possibly adding keywords to frontMatter and including them in the index

B2o5T reviewed (commented):
Thanks for your PR to improve the search.

Since your PR contains a lot of changes, unit tests are highly required, to see how search results become different.

I don't know what the proper way of testing this, I tried to help you and added initial tests for flexsearch in the following PR https://github.com/shuding/nextra/pull/2455

I don't think that want to make some changes for v2 since v3 will be out soon. Could you change your base branch to https://github.com/shuding/nextra/pull/2455 ?

I added a falling test for your case https://github.com/shuding/nextra/blob/62eb7436bfd3fec9aa7a75d0cca03127c9ae4444/packages/nextra-theme-docs/__tests__/search.test.ts#L25-L27

and for some bug that I saw previously https://github.com/shuding/nextra/blob/62eb7436bfd3fec9aa7a75d0cca03127c9ae4444/packages/nextra-theme-docs/__tests__/search.test.ts#L21-L23

Let's fill this file with unexpected search results and see how results will be changed in your PR

B2o5T commented on packages/nextra-theme-docs/src/components/flexsearch.tsx:
```diff
@@ -99,46 +87,166 @@ const loadIndexesImpl = async (
-        pageId: `page_${pageId}`,
-        content: title,
-        ...(paragraphs[0] && { display: paragraphs[0] })
+        content: content.replace(/\s+/g, ' ')
```
```suggestion
        content
```
this replace is better to add to remark plugin https://github.com/shuding/nextra/blob/62eb7436bfd3fec9aa7a75d0cca03127c9ae4444/packages/nextra/src/server/mdx-plugins/remark-structurize.ts#L26-L34

B2o5T commented on packages/nextra-theme-docs/src/components/flexsearch.tsx:
```diff
@@ -152,95 +260,154 @@ export function Flexsearch({
+    }
+
+    // get top scoring pages
+    let scoredDocs = Object.values(scoredDocMap)
+    scoredDocs.sort((a, b) => b.score - a.score)
+    scoredDocs = scoredDocs.slice(0, PAGE_LIMIT)
```
```suggestion
    const scoredDocs = Object.values(scoredDocMap)
      .sort((a, b) => b.score - a.score)
      .slice(0, PAGE_LIMIT)
```

simplify

silas commented:
> Since your PR contains a lot of changes, unit tests are highly required, to see how search results become different.

Absolutely. I'll also include unit tests for the helper functions as well.

> I don't know what the proper way of testing this, I tried to help you and added initial tests for flexsearch in the following PR #2455

Yeah, I think it will be a mix of unit tests and manual testing. Creating unit tests for "good" results is going to be tricky. But I'll definitely do a first pass of all the obvious stuff.

I did a fair bit of manual testing using my docs and the nextra.site docs, but I might also try to pull down the data files from sites listed in the showcase and see how those look.

I also need to do some testing in other languages, I've only tested English thus far.

We might also need to tweak the summary length and such based on the language (need to do a little research on Javascript unicode handling in regards to regex word boundaries and string lengths... it's been a while since I did anything non-trivial).

> I don't think that want to make some changes for v2 since v3 will be out soon. Could you change your base branch to #2455 ?

Definitely. Just timing wise, assuming this ends of being useful and you decide to merge, is it something you'd want to get in before the v3 GA release? If so I can make sure I get it done before then.

> I added a falling test for your case
> 
> https://github.com/shuding/nextra/blob/62eb7436bfd3fec9aa7a75d0cca03127c9ae4444/packages/nextra-theme-docs/__tests__/search.test.ts#L25-L27
> 
> and for some bug that I saw previously
> 
> https://github.com/shuding/nextra/blob/62eb7436bfd3fec9aa7a75d0cca03127c9ae4444/packages/nextra-theme-docs/__tests__/search.test.ts#L21-L23

Awesome.

> Let's fill this file with unexpected search results and see how results will be changed in your PR

Will do.

silas commented:
## Suspense

Before

<img width="588" alt="before" src="https://github.com/shuding/nextra/assets/18528/b5cedb86-da39-4667-9541-9cbb40738b7d">

After

<img width="596" alt="after" src="https://github.com/shuding/nextra/assets/18528/06615d68-e143-48e3-9e56-d6534a6130b5">

## This is a page

Before

<img width="590" alt="before" src="https://github.com/shuding/nextra/assets/18528/23997e5d-9aa8-4721-b3c5-8fa0c52d08d2">

After

<img width="595" alt="after" src="https://github.com/shuding/nextra/assets/18528/96137a8d-d19d-4357-9d95-5569acff0379">

## SWRConfig

Before

<img width="588" alt="before" src="https://github.com/shuding/nextra/assets/18528/4712a709-a85f-40ec-80c6-7d93cbddc1a6">

After

<img width="590" alt="after" src="https://github.com/shuding/nextra/assets/18528/19929e48-3ea8-4420-9dd5-801de2fb769b">

## Pre-fill

Before

<img width="588" alt="before" src="https://github.com/shuding/nextra/assets/18528/00b8b28b-f279-47e2-bf2a-48bc3720bee7">

After

<img width="589" alt="after" src="https://github.com/shuding/nextra/assets/18528/b048c188-325d-4004-989f-9538e2dcace4">

silas commented:
## Fourth Long ToC Item That Definitely Needs to Wrap

Before

<img width="595" alt="before" src="https://github.com/shuding/nextra/assets/18528/2a4b6566-2df8-4e09-9b30-ec52a4174084">

After

<img width="599" alt="after" src="https://github.com/shuding/nextra/assets/18528/cd9aa14a-bacc-422d-8cb1-7b3ca4c357a1">

## This page should not have the sidebar

Before

<img width="597" alt="before" src="https://github.com/shuding/nextra/assets/18528/2698c66a-bea0-4716-a137-dff2a116fc3f">

After

<img width="591" alt="after" src="https://github.com/shuding/nextra/assets/18528/6a294d46-2b3c-48e4-b4ce-84c3a9a4b413">

## google.com

<img width="599" alt="before" src="https://github.com/shuding/nextra/assets/18528/f233c0c7-8de4-4bb1-9d53-e6c0e1e8af7d">

<img width="588" alt="after" src="https://github.com/shuding/nextra/assets/18528/1d5ad60c-55e5-4fdc-a408-a4d202b7c1b3">

## Key Change + Fallback

Before

<img width="596" alt="before" src="https://github.com/shuding/nextra/assets/18528/33ac0920-25dd-4b9f-b621-7bf127d02063">

After

<img width="591" alt="after" src="https://github.com/shuding/nextra/assets/18528/bbf5b36d-03f4-43ab-9ded-3451a3bf2c77">

B2o5T commented:
@silas thank you for working on it, I am in retreat right now, will review your work next week

silas commented:
> @silas thank you for working on it, I am in retreat right now, will review your work next week

Yeah no worries, I'm just pasting screenshots so I can remember to fix issues, but I should probably just be doing that locally.

I'll mark it as ready for review and directly ping you once it's ready, didn't mean to spam your notifications.

B2o5T commented:
Thanks! Sounds good

silas commented:
@dimaMachina ready for review when your back next week

dimaMachina commented on packages/nextra-theme-docs/src/components/flexsearch.tsx:
```diff
@@ -271,8 +511,46 @@ export function Flexsearch({
+              <>
+                <div className="_text-base _font-semibold _leading-5">
+                  <Content
+                    parts={section.title.length ? section.title : page.title}
```
```suggestion
                    parts={section.title || page.title}
```

silas replied:
These are now lists, so this pattern doesn't work anymore.

silas commented:
@dimaMachina Friendly ping. Anything I can do to make this easier to review?

dimaMachina reviewed (commented):
Hey @silas, first of all, sorry for taking so long to not review and not merge your wonderful PR. I took a long vacation and am still on it.

Your work looks incredible, and I am very appreciative of it.

In the meantime, I recently discovered the https://github.com/shuding/nextra/issues/2524 proposal, and it just blew my mind:

- it removes complexity under the current search API
- it should work with remote content and data fetched from `getStaticProps` https://github.com/shuding/nextra/issues/1611 https://github.com/shuding/nextra/issues/1980
- it decreases the bundle size of nextra-theme-docs (see my experimenting PR https://github.com/shuding/nextra/pull/2569)

I just pushed my draft PR https://github.com/shuding/nextra/pull/2569 and if we find a way to fix all the missing tasks that I described, I would definitely stay with Pagefind instead of Flexsearch.

If not, we'll stay with your work done ❤️

silas commented:
@dimaMachina Cool. This definitely adds a bunch of complexity and I think it's probably better to go with a simpler solution if you can.

We decided to switch to markdoc, so I probably won't have the time/motivation to get this merged in the future.

If any the code seems useful, feel free to/copy fork as you see fit!
